src/wayland/protocols/corner_radius.rs

```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+pub trait CornerRadiusHandler {
+    fn corner_radius_state(&mut self) -> &mut CornerRadiusState;
+    fn toplevel_from_resource(&mut self, toplevel: &XdgToplevel) -> Option<ToplevelSurface>;
+    fn set_corner_radius(
+        &mut self,
+        toplevel: &cosmic_corner_radius_toplevel_v1::CosmicCornerRadiusToplevelV1,
+        data: &CornerRadiusData,
+    );
+    fn unset_corner_radius(
+        &mut self,
+        toplevel: &cosmic_corner_radius_toplevel_v1::CosmicCornerRadiusToplevelV1,
+        data: &CornerRadiusData,
+    );
+}
+
+impl<D> GlobalDispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, (), D>
+    for CornerRadiusState
+where
+    D: GlobalDispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, ()>
+        + Dispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, ()>
+        + Dispatch<cosmic_corner_radius_toplevel_v1::CosmicCornerRadiusToplevelV1, CornerRadiusData>
+        + CornerRadiusHandler
+        + 'static,
+{
+    fn bind(
+        state: &mut D,
+        _handle: &DisplayHandle,
+        _client: &Client,
+        resource: smithay::reexports::wayland_server::New<
+            cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1,
+        >,
+        _global_data: &(),
+        data_init: &mut smithay::reexports::wayland_server::DataInit<'_, D>,
+    ) {
+        let instance = data_init.init(resource, ());
+        state.corner_radius_state().instances.push(instance);
+    }
+}
+
+impl<D> Dispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, (), D>
+    for CornerRadiusState
+where
+    D: GlobalDispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, ()>
+        + Dispatch<cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1, ()>
+        + Dispatch<cosmic_corner_radius_toplevel_v1::CosmicCornerRadiusToplevelV1, CornerRadiusData>
+        + CornerRadiusHandler
+        + 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        _resource: &cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1,
+        request: <cosmic_corner_radius_manager_v1::CosmicCornerRadiusManagerV1 as smithay::reexports::wayland_server::Resource>::Request,
+        _data: &(),
+        _dhandle: &DisplayHandle,
+        data_init: &mut smithay::reexports::wayland_server::DataInit<'_, D>,
+    ) {
+        match request {
+            cosmic_corner_radius_manager_v1::Request::Destroy => {
+                let corner_radius_state = state.corner_radius_state();
+                corner_radius_state.instances.retain(|i| i != _resource);
+            }
+            cosmic_corner_radius_manager_v1::Request::GetCornerRadius { id, toplevel } => {
+                let data = Mutex::new(CornerRadiusInternal {
+                    toplevel: toplevel.downgrade(),
+                    top_left: 0,
+                    top_right: 0,
+                    bottom_right: 0,
+                    bottom_left: 0,
```

Comment:
So this means any toplevel, that has a corner-radius object gets no rounded outline, if the client never sets radius values?

That would be incorrect, we want to apply the same behaviour here, as if the client didn't use the protocol. Same after `unset_radius` btw until it sets a new one.

Which means this needs to be an `Option` probably.